vertx-grpc-protoc-plugin2/src/main/java/io/vertx/grpc/plugin/generation/generators/CodeGenerator.java

```diff
@@ -0,0 +1,43 @@
+package io.vertx.grpc.plugin.generation.generators;
+
+import io.vertx.grpc.plugin.generation.*;
+import io.vertx.grpc.plugin.template.TemplateEngine;
+import io.vertx.grpc.plugin.template.TemplateException;
+
+/**
+ * Abstract base class for a code generator that produces generated files based on a specific type of generation context.
+ * <p>
+ * This class provides common functionalities such as file creation, relative path calculation, and integration with a template engine to facilitate rendering templates into
+ * generated files. Concrete implementations should extend this class and provide specific logic for the code generation process.
+ */
+public abstract class CodeGenerator {
+
+  protected final TemplateEngine templateEngine;
```

Comment:
the CodeGenerator does not need to keep a reference to the template engine as we only have a single engine that is the mustache engine, this will simplify a bit.